File: src/java/foundgine-core/src/main/java/com/foundgine/core/execution/mutation/IMutationExecutionProvider.java
package com.foundgine.core.execution.mutation;

import com.foundgine.core.execution.CancellationToken;
import com.foundgine.core.execution.ExecutionContext;

/** Port of {@code Foundgine.Core.Execution.Mutation.IMutationExecutionProvider}. */
public interface IMutationExecutionProvider {

    MutationResult execute(ProviderMutationPlan plan, ExecutionContext context);

    /**
     * Cancellation-aware execution boundary. The default method mirrors the
     * C# default-interface-method overload; see {@link CancellationToken}
     * for the caveat that Java's cancellation model here is a placeholder.
     */
    default MutationResult execute(ProviderMutationPlan plan, ExecutionContext context, CancellationToken cancellationToken) {
        cancellationToken.throwIfCancellationRequested();
        return execute(plan, context);
    }
}

File: src/csharp/Foundgine.Core/Execution/Mutation/IMutationExecutionProvider.cs
namespace Foundgine.Core.Execution.Mutation;

public interface IMutationExecutionProvider
{
    MutationResult Execute(
        ProviderMutationPlan plan,
        ExecutionContext context);

    MutationResult Execute(
        ProviderMutationPlan plan,
        ExecutionContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        return Execute(plan, context);
    }
}
